cloud/packet/public/resources/device.go

```diff
@@ -0,0 +1,368 @@
+	var device *packngo.Device
+	for j := actualResource.Count; j < expectedResource.Count; j++ {
+		hostname := fmt.Sprintf("%s-%d", expected.(*Device).Name, j)
+		createRequest := &packngo.DeviceCreateRequest{
+			HostName:     hostname,
+			Facility:     expectedResource.Location,
+			Plan:         expectedResource.Type,
+			OS:           expectedResource.OS,
+			UserData:     string(userData),
+			Tags:         expectedResource.Tags,
+			ProjectID:    projectID,
+			BillingCycle: "hourly",
+		}
+		logger.Debug("creating server %s: %v", hostname, createRequest)
+		device, _, err = Sdk.Client.Devices.Create(createRequest)
+		if err != nil {
+			return nil, nil, err
+		}
+		logger.Success("Created Device [%s]", device.Hostname)
+	}
+
+	// if these are masters, we are not done until we have the master IP
+	if r.ServerPool.Type == cluster.ServerPoolTypeMaster {
+		masterIPs, err := getMasterIP(projectID, cluster.ServerPoolTypeMaster)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		found := masterIPs[0] != "" && masterIPs[1] != "" && masterIPs[2] != ""
+		logger.Debug("device.Apply master IP addresses %t %s %s %s", found, masterIPs[0], masterIPs[1], masterIPs[2])
+
+		if !found {
+			return nil, nil, fmt.Errorf("Unable to find master IP addresses")
+		}
+
+		immutable.KubernetesAPI.Endpoint = masterIPs[0]
+
+	}
+
+	newResource := &Device{
+		Shared: Shared{
+			Name: r.ServerPool.Name,
+		},
+		BootstrapScripts: expected.(*Device).BootstrapScripts,
+	}
+	if expectedResource.Count != 0 && device != nil {
+		newResource.OS = device.OS.Slug
+		newResource.Type = device.Plan.Slug
+		newResource.Location = device.Facility.Code
+		newResource.Count = r.Count
+	}
+
+	logger.Debug("device.Apply newResource %v", newResource)
+
+	newCluster := r.immutableRender(newResource, immutable)
+	return newCluster, newResource, nil
+}
+
+func (r *Device) Delete(actual cloud.Resource, immutable *cluster.Cluster) (*cluster.Cluster, cloud.Resource, error) {
+	logger.Debug("device.Delete")
+	deleteResource := actual.(*Device)
+	logger.Debug("device.Delete deleteResource %v", deleteResource)
+	if deleteResource.Name == "" {
+		return nil, nil, fmt.Errorf("Unable to delete device resource without Name [%s]", deleteResource.Name)
+	}
+
+	valid, err := devicesByTag(r.ProjectID, r.ServerPool.Type)
+	logger.Debug("device.Delete devices to delete %v", valid)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for _, device := range valid {
+
+		_, err := Sdk.Client.Devices.Delete(device.ID)
+		if err != nil {
+			return nil, nil, err
+		}
+		logger.Success("Deleted Device [%s]", device.Hostname)
+	}
+
+	// Kubernetes API
+	// todo (@kris-nova) this is obviously not immutable
+	immutable.KubernetesAPI.Endpoint = ""
+
+	newResource := &Device{
+		OS:               deleteResource.OS,
+		Type:             deleteResource.Type,
+		Location:         deleteResource.Location,
+		BootstrapScripts: deleteResource.BootstrapScripts,
+		Count:            0,
```

Comment:
Is `count: 0` indeed to be zero?

Reply:
I think so, but am not sure. This is at the end of a successful `Delete()` operation. At this point, is the returned `Resource` expected to return:

* the _actual_ state of the cluster, In which case it should be `0`?
* the _configured_ state of the cluster, in which case it should be what was configured originally, but we happen to have `0` because they were deleted, in which case it should be `1` (or whatever was in the profile)?

Which one should be returned here?

Reply:
Not really sure, but as far as I know, `Delete` returns actual state. Actual state is `Count: 0`, so it should stay zero. Thanks for checking with me.